### PR TITLE
i#6758: Fix AArch64 FP state at fcache events and attach

### LINKED ISSUE #6758: linux.signalNNN1 test failing due to non-determinism in floating point operations
Splitting out from #6755 since we're not sure how much they overlap.

In downstream systems we're seeing that the linux.signalNNN1 tests are failing due to non-determinism in the floating point result of some trig math ops. This for some reason does not affect the test run on Github Actions.

These failures started since https://github.com/DynamoRIO/dynamorio/pull/6725.

### PR DESCRIPTION
**Pull request by abhinav92003**
Fixes the slot used to save and restore FP regs at fcache enter and return events. PR #6725 adjusted the slots used during signal handling in core/unix/signal_linux_aarch64.c but did not adjust the same in fcache enter/return and attach events. Prior to that PR, the FP regs were simply stored in a contiguous manner in signal handling code and fcache enter/return routines (instead of in their respective dr_simd_t struct member), which was a bit confusing.

The mismatch between slot usage in signal handling and fcache enter/return code caused failures in the signalNNN1 tests on some systems. Verified that those tests pass with this fix.

Also fixes the same issue in save_priv_mcontext_helper which is used in the dr_app_start API. Unit tests for this scenario will be added as part of #6759.

Issue: #5036, #6755, #5365
Fixes #6758

### CODE REVIEW

File: core/arch/aarch64/aarch64.asm
Comment (abhinav92003) [inline core/arch/aarch64/aarch64.asm:230]: @derekbruening Do you know if any existing test provides coverage for this change? Probably not because it didn't fail on/after #6725 (but then so didn't signalNNN1 on Github Actions). Maybe we need to port detach_state to AArch64 and add more state verifications to it?

**Comment by abhinav92003** [comment]
This PR is probably not ready completely yet, but would be good to get some quick feedback since this is blocking downstream pulls.

File: core/arch/aarch64/emit_utils.c
Comment (derekbruening) [reply core/arch/aarch64/emit_utils.c:799]: DR supports a max SVE vec len of 512-bit which is 64 bytes (see comment above dr_simd_t def).

File: core/arch/aarch64/aarch64.asm
Comment (derekbruening) [inline core/arch/aarch64/aarch64.asm:1]: I would file a separate dedicated bug on the signal test failures we are seeing: I think that will be easier to understand and track this fix and have something to point at for what is blocking us.

Comment (derekbruening) [inline core/arch/aarch64/aarch64.asm:1]: > to that PR the FP regs were simply stored in a continuous manner in signal handling code, and fcache enter/exit routines 

s/continuous/contiguous/
s/code, and/code and/

File: core/arch/aarch64/aarch64.asm
Comment (derekbruening) [reply core/arch/aarch64/aarch64.asm:230]: client.cleancall has some mcontext field checks: but it looks like it was never ported to aarchxxx.  Yes, getting more testing parity on aarchxx would be good.

File: core/arch/aarch64/emit_utils.c
Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c]: Ah thanks for clarifying. I made the changes to the sve code too. PTAL.

File: core/arch/aarch64/aarch64.asm
Comment (abhinav92003) [reply core/arch/aarch64/aarch64.asm:1]: Filed #6758.

Comment (abhinav92003) [reply core/arch/aarch64/aarch64.asm:1]: Done.

File: core/arch/aarch64/aarch64.asm
Comment (abhinav92003) [reply core/arch/aarch64/aarch64.asm:230]: But let's not block this PR on that I assume?

Comment (derekbruening) [reply core/arch/aarch64/aarch64.asm:230]: Agreed.

File: core/arch/aarch64/aarch64.asm
Comment (derekbruening) [inline core/arch/aarch64/aarch64.asm:231]: In addition to assuming the vec len is the full 64 bytes, was the ST1 also storing just one D (half-size element) instead of the whole V==Q?

File: core/arch/aarch64/emit_utils.c
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c:607]: Wait, are the pred regs stored in a dr_simd_t?  Indeed they are in mcxt_api.h: but aren't they 1/8 the size?  Isn't that a waste of space?  @AssadHashmi could you help us understand that?

**Review by derekbruening — APPROVED** [review]
The change makes sense to me but probably best to have @AssadHashmi take a look as I am not an expert in a64 SIMD: as you can tell I'm confused by the svep field being a full dr_simd_t and other things.

File: core/arch/aarch64/emit_utils.c
Comment (jackgallagher-arm) [inline core/arch/aarch64/emit_utils.c:607]: Lines `634` and `848` also load `P15` as part of saving/restoring the `FFR` register so those need to be updated too.

`/15 * (proc_get_vector_length_bytes() / 8)/15 * sizeof(dr_simd_t)/`

I think this is what is causing the `P15` corruption errors in the scatter/gather test.

**Review by jackgallagher-arm — CHANGES_REQUESTED** [review]

File: core/arch/aarch64/emit_utils.c
Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: Done.

Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: PTAL

**Review by jackgallagher-arm — APPROVED** [review]

File: core/arch/aarch64/aarch64.asm
Comment (abhinav92003) [reply core/arch/aarch64/aarch64.asm:231]: As I understand the previous version did vx.2d which means 2 64-bit floating point = 128 bits. And it stored 4 such together (==64 bytes) in one st1 op. @jackgallagher-arm to confirm.

File: core/arch/aarch64/emit_utils.c
Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: @jackgallagher-arm any thoughts on this?

File: core/arch/aarch64/aarch64.asm
Comment (jackgallagher-arm) [reply core/arch/aarch64/aarch64.asm:231]: > As I understand the previous version did vx.2d which means 2 64-bit floating point = 128 bits. And it stored 4 such together (==64 bytes) in one st1 op. @jackgallagher-arm to confirm

That is correct.

File: core/arch/aarch64/emit_utils.c
Comment (AssadHashmi) [reply core/arch/aarch64/emit_utils.c:607]: Yes, predicates and FFR are 1/8th of the size. The idea was to get SVE working first then make such refinements, and this was a quick way of getting things up and running. Should we raise an issue to fix this or do you want to do it as part of this patch?

Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: Separate issue would be better, since this PR is needed to unblock our downstream merge.

File: core/arch/aarch64/aarch64.asm
Comment (AssadHashmi) [reply core/arch/aarch64/aarch64.asm:230]: I've created https://github.com/DynamoRIO/dynamorio/issues/6759 to improve coverage. If you have any other suggestions of x86 tests not ported to AArch64 which will catch bugs like this, please add them to that issue.

**Comment by abhinav92003** [comment]
code_api|tool.drcov.eintr failed on ubuntu-20-arm64-sve. I see it did so also on #6756 though so very likely unrelated to this PR itself.

x86-32 tool.drcachesim.opcode_mix failure is #6303.

File: core/arch/aarch64/emit_utils.c
Comment (AssadHashmi) [reply core/arch/aarch64/emit_utils.c:607]: Created https://github.com/DynamoRIO/dynamorio/issues/6760.

**Review by AssadHashmi — APPROVED** [review]

**Comment by abhinav92003** [comment]
#6761 is fixed now. Rerunning tests.

File: core/arch/aarch64/emit_utils.c
Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: Thanks!

File: core/arch/aarch64/aarch64.asm
Comment (abhinav92003) [reply core/arch/aarch64/aarch64.asm:230]: Thanks for creating #6759!

File: core/arch/aarch64/emit_utils.c
Comment (derekbruening) [reply core/arch/aarch64/emit_utils.c:607]: Note that such a change in the mcontext layout would break compatibility and require a major version change.

File: core/arch/aarch64/aarch64.asm
Comment (derekbruening) [reply core/arch/aarch64/aarch64.asm:231]: OK, so there are 2 variants of `ST1`: single-lane and multi-lane, all under the same `ST1` opcode.  But why call it `ST1` for multi-lane?  I thought the `1` meant "one lane".

File: core/arch/aarch64/emit_utils.c
Comment (derekbruening) [reply core/arch/aarch64/emit_utils.c:607]: Oh you mean PR #6725 changes the size of dr_simd_t!  That didn't sink in when I reviewed it.  Yes, it breaks compatibility and it should have bumped the minor version to 90 as part of the same PR per https://dynamorio.org/page_new_release.html#autotoc_md230 and clearly documented the breakage in the release notes.  @AssadHashmi please submit a follow-up PR for that.

Comment (derekbruening) [reply core/arch/aarch64/emit_utils.c:607]: Wait, PR #6725 doesn't change the size: the new union field member is the same 64 bytes.

@abhinav92003 please clarify what you mean about breaking compatibility

Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: PR #6725 doesn't break compatibility. But it sets the data differently in the existing field. I assume that also needs to be backward compatible (not just the struct definition).

Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: Changing how we set data is set in the existing same-size field also would require clients to be updated. They would need to read the struct accordingly.

Comment (derekbruening) [reply core/arch/aarch64/emit_utils.c:607]: Before, the SIMD data was written in a way that didn't even line up with the dr_simd_t structs, right?  Someone trying to manually pull the data out would be very confused.  In fact: was there a users list question about that with someone not able to get the data?  This seems like a bug fix.  Maybe someone figured out the way it was written before and wrote client code assuming that but it seems like it's not the same kind of change where we broke a documented interface.

Comment (abhinav92003) [reply core/arch/aarch64/emit_utils.c:607]: Right. That's what I meant.

> Though it really fixes it by spilling to the slots one would expect

**Pull request merged**
